In [1]:
import numpy as np

import pandas as pd

In [2]:
# Read CSV file given as an Input from the user. Here, I took index column as ID

df = pd.read_csv("chrome_reviews.csv",index_col='ID')

df

Review URL  \
ID                                                          
3886    https://play.google.com/store/apps/details?id=...   
3887    https://play.google.com/store/apps/details?id=...   
3888    https://play.google.com/store/apps/details?id=...   
3889    https://play.google.com/store/apps/details?id=...   
3890    https://play.google.com/store/apps/details?id=...   
...                                                   ...   
684987  https://play.google.com/store/apps/details?id=...   
684988  https://play.google.com/store/apps/details?id=...   
684989  https://play.google.com/store/apps/details?id=...   
684990  https://play.google.com/store/apps/details?id=...   
684991  https://play.google.com/store/apps/details?id=...   

                                                     Text  Star  Thumbs Up  \
ID                                                                           
3886                           This is very helpfull aap.     5          0   
3887                                                 Good     3          2   
3888       Not able to update. Neither able to uninstall.     1          0   
3889                                             Nice app     4          0   
3890                                    Many unwanted ads     1          0   
...                                                   ...   ...        ...   
684987                                          Bagusss..     5          0   
684988                                      Bad version 😔     1          0   
684989  One thing that I have to say I can't spelled t...     5          0   
684990                                          Excellent     5          0   
684991  After update it lag and always slow same goes ...     1          0   

               User Name Developer Reply        Version Review Date  \
ID                                                                    
3886    INDIAN Knowledge             NaN  83.0.4103.106  2020-12-19   
3887    Ijeoma Happiness             NaN  85.0.4183.127  2020-12-19   
3888    Priti D BtCFs-29             NaN  85.0.4183.127  2020-12-19   
3889          Ajeet Raja             NaN  77.0.3865.116  2020-12-19   
3890             Rams Mp             NaN   87.0.4280.66  2020-12-19   
...                  ...             ...            ...         ...   
684987   boima panjaitan             NaN   88.0.4324.93  2021-03-05   
684988      निशान्त सिंह             NaN  88.0.4324.181  2021-03-05   
684989      Virgie Allen             NaN   89.0.4389.72  2021-03-05   
684990    Kazuo Guevarra             NaN   89.0.4389.72  2021-03-05   
684991        Hanin Amni             NaN   88.0.4324.93  2021-03-05   

                    App ID  
ID                          
3886    com.android.chrome  
3887    com.android.chrome  
3888    com.android.chrome  
3889    com.android.chrome  
3890    com.android.chrome  
...                    ...  
684987  com.android.chrome  
684988  com.android.chrome  
684989  com.android.chrome  
684990  com.android.chrome  
684991  com.android.chrome  

[7204 rows x 9 columns]

# EDA Analysis

In [3]:
df.shape

(7204, 9)

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 7204 entries, 3886 to 684991
Data columns (total 9 columns):
 #   Column           Non-Null Count  Dtype 
---  ------           --------------  ----- 
 0   Review URL       7204 non-null   object
 1   Text             7203 non-null   object
 2   Star             7204 non-null   int64 
 3   Thumbs Up        7204 non-null   int64 
 4   User Name        7204 non-null   object
 5   Developer Reply  95 non-null     object
 6   Version          7119 non-null   object
 7   Review Date      7204 non-null   object
 8   App ID           7204 non-null   object
dtypes: int64(2), object(7)
memory usage: 562.8+ KB


#### - We can see that all the columns are non-null except Text,Developer Reply and Version.
     
#### - But we can see Some columns are not relevant to our analysis.So, we will drop those columns.

In [5]:
df.drop(['User Name','Developer Reply','Thumbs Up','Version','Review Date','App ID','Review URL'] ,axis = 1 ,inplace = True)


In [6]:
df

Text  Star
ID                                                             
3886                           This is very helpfull aap.     5
3887                                                 Good     3
3888       Not able to update. Neither able to uninstall.     1
3889                                             Nice app     4
3890                                    Many unwanted ads     1
...                                                   ...   ...
684987                                          Bagusss..     5
684988                                      Bad version 😔     1
684989  One thing that I have to say I can't spelled t...     5
684990                                          Excellent     5
684991  After update it lag and always slow same goes ...     1

[7204 rows x 2 columns]

### Check for any blank row and if it is there then we will drop it.

In [7]:
df.dropna(inplace=True)

blanks = []  # start with an empty list

for i,lb,rv in df.itertuples():  # iterate over the DataFrame
    if type(rv)==str:            # avoid NaN values
        if rv.isspace():         # test 'review' for whitespace
            blanks.append(i)     # add matching index numbers to the list

df.drop(blanks, inplace=True)

### Final filtered data for processing

In [8]:
df

Text  Star
ID                                                             
3886                           This is very helpfull aap.     5
3887                                                 Good     3
3888       Not able to update. Neither able to uninstall.     1
3889                                             Nice app     4
3890                                    Many unwanted ads     1
...                                                   ...   ...
684987                                          Bagusss..     5
684988                                      Bad version 😔     1
684989  One thing that I have to say I can't spelled t...     5
684990                                          Excellent     5
684991  After update it lag and always slow same goes ...     1

[7203 rows x 2 columns]

### Importing libraries for sentiment analysis to understand the review.

In [9]:
from nltk.sentiment.vader import SentimentIntensityAnalyzer

sid = SentimentIntensityAnalyzer()

### Assigning scores for each review

In [10]:
df['scores'] = df['Text'].apply(lambda review: sid.polarity_scores(review))

df['compound']  = df['scores'].apply(lambda score_dict: score_dict['compound'])

df['comp_score'] = df['compound'].apply(lambda c: 'pos' if c >=0 else 'neg')

df

Text  Star  \
ID                                                                
3886                           This is very helpfull aap.     5   
3887                                                 Good     3   
3888       Not able to update. Neither able to uninstall.     1   
3889                                             Nice app     4   
3890                                    Many unwanted ads     1   
...                                                   ...   ...   
684987                                          Bagusss..     5   
684988                                      Bad version 😔     1   
684989  One thing that I have to say I can't spelled t...     5   
684990                                          Excellent     5   
684991  After update it lag and always slow same goes ...     1   

                                                   scores  compound comp_score  
ID                                                                              
3886    {'neg': 0.0, 'neu': 1.0, 'pos': 0.0, 'compound...    0.0000        pos  
3887    {'neg': 0.0, 'neu': 0.0, 'pos': 1.0, 'compound...    0.4404        pos  
3888    {'neg': 0.0, 'neu': 1.0, 'pos': 0.0, 'compound...    0.0000        pos  
3889    {'neg': 0.0, 'neu': 0.263, 'pos': 0.737, 'comp...    0.4215        pos  
3890    {'neg': 0.487, 'neu': 0.513, 'pos': 0.0, 'comp...   -0.2263        neg  
...                                                   ...       ...        ...  
684987  {'neg': 0.0, 'neu': 1.0, 'pos': 0.0, 'compound...    0.0000        pos  
684988  {'neg': 0.778, 'neu': 0.222, 'pos': 0.0, 'comp...   -0.5423        neg  
684989  {'neg': 0.06, 'neu': 0.634, 'pos': 0.306, 'com...    0.8395        pos  
684990  {'neg': 0.0, 'neu': 0.0, 'pos': 1.0, 'compound...    0.5719        pos  
684991  {'neg': 0.136, 'neu': 0.809, 'pos': 0.055, 'co...   -0.4055        neg  

[7203 rows x 5 columns]

### Filtering out worst rated movie on the basis of its compound score

### Note: Here, our aim is to take out those movies where comp_score is positive but still star is 1.



In [11]:
df_worst_rated = df[(df.loc[:,'Star'] == 1) & (df.loc[:,'comp_score'] == 'pos') & (df.loc[:,'compound'] >0)]

### Filtered movies where review does not justify the ratings

In [12]:
df_worst_rated

Text  Star  \
ID                                                                
3898    App is not getting update and it is not gettin...     1   
3901                Very coming in real status. Thank you     1   
3905    I want to download greana free fire but that n...     1   
3925    Your download manager is garbage. Pls fix 😠 ne...     1   
3928                          Okk kind but bad then brave     1   
...                                                   ...   ...   
659656                                           Good app     1   
659661                                           Its good     1   
659680  Im not able to update chrome on my mobile plea...     1   
684939                            No good its not updated     1   
684983  Chrome is good, I trust this update will make ...     1   

                                                   scores  compound comp_score  
ID                                                                              
3898    {'neg': 0.0, 'neu': 0.949, 'pos': 0.051, 'comp...    0.0726        pos  
3901    {'neg': 0.0, 'neu': 0.706, 'pos': 0.294, 'comp...    0.3612        pos  
3905    {'neg': 0.121, 'neu': 0.643, 'pos': 0.236, 'co...    0.1531        pos  
3925    {'neg': 0.0, 'neu': 0.86, 'pos': 0.14, 'compou...    0.0772        pos  
3928    {'neg': 0.326, 'neu': 0.206, 'pos': 0.467, 'co...    0.2617        pos  
...                                                   ...       ...        ...  
659656  {'neg': 0.0, 'neu': 0.256, 'pos': 0.744, 'comp...    0.4404        pos  
659661  {'neg': 0.0, 'neu': 0.256, 'pos': 0.744, 'comp...    0.4404        pos  
659680  {'neg': 0.0, 'neu': 0.839, 'pos': 0.161, 'comp...    0.3182        pos  
684939  {'neg': 0.272, 'neu': 0.37, 'pos': 0.358, 'com...    0.1779        pos  
684983  {'neg': 0.0, 'neu': 0.435, 'pos': 0.565, 'comp...    0.8442        pos  

[382 rows x 5 columns]

### So, there are total 382 such users ID who gave incorrect rating

In [13]:
df_worst_rated.shape

(382, 5)

In [14]:
df_incorrect_reviews = df_worst_rated.drop(['scores','comp_score','compound'],axis= 1)

In [15]:
df_incorrect_reviews

Text  Star
ID                                                             
3898    App is not getting update and it is not gettin...     1
3901                Very coming in real status. Thank you     1
3905    I want to download greana free fire but that n...     1
3925    Your download manager is garbage. Pls fix 😠 ne...     1
3928                          Okk kind but bad then brave     1
...                                                   ...   ...
659656                                           Good app     1
659661                                           Its good     1
659680  Im not able to update chrome on my mobile plea...     1
684939                            No good its not updated     1
684983  Chrome is good, I trust this update will make ...     1

[382 rows x 2 columns]

In [16]:
df_incorrect_reviews.shape

(382, 2)

In [18]:
print(df_incorrect_reviews['Text'].groupby(df['Star']).value_counts())

Star  Text                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                
1     Good                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                   

In [19]:
df_incorrect_reviews

Text  Star
ID                                                             
3898    App is not getting update and it is not gettin...     1
3901                Very coming in real status. Thank you     1
3905    I want to download greana free fire but that n...     1
3925    Your download manager is garbage. Pls fix 😠 ne...     1
3928                          Okk kind but bad then brave     1
...                                                   ...   ...
659656                                           Good app     1
659661                                           Its good     1
659680  Im not able to update chrome on my mobile plea...     1
684939                            No good its not updated     1
684983  Chrome is good, I trust this update will make ...     1

[382 rows x 2 columns]

In [37]:
df_incorrect_reviews.to_csv('Cleaned_reviews.csv')